### PR TITLE
Denial Constraints Tutorial 1

### PR DESCRIPTION
**Pull request by jw-mcgrath**

### CODE REVIEW

**Comment by jw-mcgrath** [comment]
Need to add template for error detection modules

**Review by j48zheng — CHANGES_REQUESTED** [review]
Move the table of Denial Constraint Operators up since they are used earlier in the tutorial.
As an additional comment: should we include the non-symmetrical operators in the table for this release, since holoclean does not support it properlly at the moment.

**Comment by jw-mcgrath** [comment]
Hey all, I have made some changes to make sure everything works correctly, and made some slight changes to wording. @thodrek, could you comment on the order in which I introduce the operators?

File: tutorial/Denial Constraints.ipynb [hunk 1]
```diff
@@ -0,0 +1,198 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Introduction to Denial Constraints\n",
+    "\n",
+    "HoloClean's goal is to clean your data, and the system is driven by a description of what clean data *should not* be like. These are expressed in the form of a Denial Constraint, which is similar to a functional dependency. However functional dependencies express things that should hold for your data, a denial constraint expresses what clean data is not like.\n",
```
Comment (thodrek) [inline tutorial/Denial Constraints.ipynb]: "what clean data should not be like. " -> "what clean data should look like" we are asking people to specify invariances that hold. "which is similar to a functional dependency" -> add a hyperlink to the wikipedia entry of functional dependency

File: tutorial/Denial Constraints.ipynb [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    "Score,\n",
+    "Sample,\n",
+    "Stateavg`\n",
+    "\n",
+    "\n",
+    "And we know that there are some errors in our data. For example some people have mistyped the city name, and so we see results like\n",
+    "\n",
+    "|City|ZipCode|\n",
+    "|----|-------|\n",
+    "|Chicago|60607|\n",
```
Comment (thodrek) [inline tutorial/Denial Constraints.ipynb]: Is Chicago present in the hospital dataset? I don't think so. change it to contain examples from the actual dataset. Even better load and query the dataset while you point them to the loading tutorial.

File: tutorial/Denial Constraints.ipynb [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    "|`GT(x.y, z.w)`| `x.y > z.y`|\n",
+    "|`GTE(x.y, z.w)`| `x.y >= z.y`|\n",
+    "|`LT(x.y, z.w)`| `x.y < z.y`|\n",
+    "|`LT(x.y, z.w)`| `x.y <= z.y`|\n",
+    "\n",
+    "All denial constraints are of the form `t1&t2&<X>&<Y>&...`  where `<X>` and `<Y>` are logical operators mentioned above.\n",
+    "\n",
+    "# Next Steps\n",
+    "\n",
+    "Denial Constraints are just one of HoloClean's error detectors that it uses for learning, if you'd like to write your own check out our error detectors tutorial. If you want to learn about the next steps in the HoloClean pipeline, check out the repairs tutorial."
```
Comment (thodrek) [inline tutorial/Denial Constraints.ipynb]: Add a hyperlink to the repairs tutorial.

**Review by thodrek — CHANGES_REQUESTED** [review]

**Review by gmichalo — DISMISSED** [review]
When I run the notebook I got this error:
DCFormatException: Duplicate Denial Constraint

In the line
session.add_denial_constraint()
we need to add a dc that we do not already have

**Comment by jw-mcgrath** [comment]
I fixed these changes, please review again.

**Review by gmichalo — COMMENTED** [review]
I think it is better that the first cell in the notebook should be the title and the description and afterwards we can write the loading of the holoclean session with a small description in three cells : 
cell 1)Firstly we create a holoclean session and we ingest the Hospital dataset. 
cell 2) The code that we have
cell 3) The data has the following fields: ...

**Review by thodrek — APPROVED** [review]
Looks good to me now.

**Review by j48zheng — APPROVED** [review]

**Review by gmichalo — APPROVED** [review]

**Pull request merged**